rs-matter-macros-impl/src/tlv.rs

```diff
@@ -0,0 +1,511 @@
+           impl #generics #krate::tlv::FromTLV <#lifetime> for #enum_name #generics {
+               fn from_tlv(t: &#krate::tlv::TLVElement<#lifetime>) -> Result<Self, #krate::error::Error> {
+                   let mut t_iter = t.confirm_struct()?.enter().ok_or_else(|| #krate::error::Error::new(#krate::error::ErrorCode::Invalid))?;
+                   let mut item = t_iter.next().ok_or_else(|| Error::new(#krate::error::ErrorCode::Invalid))?;
+                   if let TagType::Context(tag) = item.get_tag() {
+                       match tag {
+                           #(
+                               #tags => Ok(Self::#variant_names(#types::from_tlv(&item)?)),
+                           )*
+                           _ => Err(#krate::error::Error::new(#krate::error::ErrorCode::Invalid)),
+                       }
+                   } else {
+                       Err(#krate::error::Error::new(#krate::error::ErrorCode::TLVTypeMismatch))
+                   }
+               }
+           }
+    };
+
+    //        panic!("Expanded to {}", expanded);
+    expanded
+}
+
+/// Derive FromTLV Macro
+///
+/// This macro works for structures. It will create an implementation
+/// of the FromTLV trait for that structure.  All the members of the
+/// structure, sequentially, will get Context tags starting from 0
+/// Some configurations are possible through the 'tlvargs' attributes.
+/// For example:
+///  #[tlvargs(lifetime = "'a", start = 1, datatype = "list", unordered)]
+///
+/// start: This can be used to override the default tag from which the
+///        decoding starts (Default: 0)
+/// datatype: This can be used to define whether this data structure is
+///        to be decoded as a structure or list. Possible values: list
+///        (Default: struct)
+/// lifetime: If the structure has a lifetime annotation, use this variable
+///        to indicate that. The 'impl' will then use that lifetime
+///        indicator.
+/// unordered: By default, the decoder expects that the tags are in
+///        sequentially increasing order. Set this if that is not the case.
+///
+/// Additionally, structure members can use the tagval attribute to
+/// define a specific tag to be used
+/// For example:
+///  #[tagval(22)]
+///  name: u8,
+/// In the above case, the 'name' attribute will be encoded/decoded with
+/// the tag 22
+pub fn derive_fromtlv(ast: DeriveInput, crate_name: String) -> TokenStream {
+    let name = &ast.ident;
+
+    let tlvargs = parse_tlvargs(&ast, crate_name);
+
+    let generics = ast.generics;
+
+    if let syn::Data::Struct(syn::DataStruct {
+        fields: syn::Fields::Named(ref fields),
+        ..
+    }) = ast.data
+    {
+        gen_fromtlv_for_struct(fields, name, tlvargs, &generics)
+    } else if let syn::Data::Enum(data_enum) = ast.data {
+        gen_fromtlv_for_enum(&data_enum, name, tlvargs, &generics)
+    } else {
+        panic!(
+            "Derive FromTLV - Only supported Struct for now {:?}",
```

Comment:
This panic is a lie - we support enum too. However I mostly only copied&pasted things.